lib/WordDiff.ml

```diff
@@ -43,98 +77,67 @@ let edit_distance (type a) (compare : a -> a -> bool) (s : a array)
+let is_approximately_equal s1 s2 =
+  let compare_char c1 c2 = c1 = c2 in
+  let s1_array = Array.of_seq (String.to_seq s1) in
+  let s2_array = Array.of_seq (String.to_seq s2) in
+  let threshold = max 3 (max (String.length s1) (String.length s2) / 3) in
+  edit_distance compare_char s1_array s2_array <= threshold
+
+let lacs words1 words2 =
+  let m = List.length words1 and n = List.length words2 in
+  let dp = Array.make_matrix (m + 1) (n + 1) [] in
+  for i = 1 to m do
+    for j = 1 to n do
+      if
+        is_approximately_equal
+          (List.nth words1 (i - 1))
+          (List.nth words2 (j - 1))
+      then dp.(i).(j) <- List.nth words1 (i - 1) :: dp.(i - 1).(j - 1)
+      else if List.length dp.(i - 1).(j) > List.length dp.(i).(j - 1) then
+        dp.(i).(j) <- dp.(i - 1).(j)
+      else dp.(i).(j) <- dp.(i).(j - 1)
+    done
+  done;
+  List.rev dp.(m).(n)
 
-let diff_words (s1 : string) (s2 : string) : line_content * line_content =
-  let words1 = Array.to_list (string_to_words s1) in
-  let words2 = Array.to_list (string_to_words s2) in
-  let common = lcs words1 words2 in
+let pair_lines (s1 : string array) (s2 : string array) :
+    (string option * string option) list =
+  let words1 = Array.to_list s1 in
+  let words2 = Array.to_list s2 in
+  let common = lacs words1 words2 in
 
-  let rec construct_diff w1 w2 lcs acc_mine acc_their =
+  let rec construct_pairs w1 w2 lcs acc =
     match (w1, w2, lcs) with
-    | [], [], [] -> (List.rev acc_mine, List.rev acc_their)
+    | [], [], [] -> List.rev acc
     | x :: xs, y :: ys, z :: zs -> (
-        match (x = z, y = z) with
+        match (is_approximately_equal x z, is_approximately_equal y z) with
         | true, true ->
-            construct_diff xs ys zs (Unchanged x :: acc_mine)
-              (Unchanged y :: acc_their)
+            construct_pairs xs ys zs ((Some x, Some y) :: acc)
         | false, true ->
-            construct_diff xs (y :: ys) (z :: zs) (Changed x :: acc_mine)
-              acc_their
+            construct_pairs xs (y :: ys) (z :: zs) ((Some x, None) :: acc)
         | true, false ->
-            construct_diff (x :: xs) ys (z :: zs) acc_mine
-              (Changed y :: acc_their)
+            construct_pairs (x :: xs) ys (z :: zs) ((None, Some y) :: acc)
         | false, false ->
-            construct_diff xs ys (z :: zs) (Changed x :: acc_mine)
-              (Changed y :: acc_their))
+            construct_pairs xs ys (z :: zs) ((Some x, Some y) :: acc))
     | x :: xs, [], lcs ->
-        construct_diff xs [] lcs (Changed x :: acc_mine) acc_their
+        construct_pairs xs [] lcs ((Some x, None) :: acc)
     | [], y :: ys, lcs ->
-        construct_diff [] ys lcs acc_mine (Changed y :: acc_their)
+        construct_pairs [] ys lcs ((None, Some y) :: acc)
     | x :: xs, y :: ys, [] ->
-        construct_diff xs ys [] (Changed x :: acc_mine) (Changed y :: acc_their)
+        construct_pairs xs ys [] ((Some x, Some y) :: acc)
```

Comment:
Pairing these lines is wrong!

Reply:
Hi, do you mind elaborating more on why it is wrong?

Reply:
Try it on the following test!
```diff
- aaaaaaaaa a
- egdg xsd fi sdgfyu s gd csue c e yeg scy
+ aaaaaaaaa b
+ yugyu fft yu utf ytfutf uyg gygu yftuf gg uf ft
```

Reply:
LACS -> A sequence of lines 
Here, we have two lines in 1 and two lines in 2. 
So LCAS is a matter of lines. Here, it will be a length of 1, and the LCAS are (Some aaaaaaaaa a, Some aaaaaaaaa b)

So here, in the accumulator, I am adding x and y as being paired, when x and y are not part of the LACS. 
Here, the example, x and y corresponds to egdg ... and yugyu ...
So they should not be paired.